### ISSUE TITLE
Wrong mapping when using long Id

### ISSUE THREAD

**Issue by manuel-ornato** [SlapperAutoMapper/Slapper.AutoMapper#48]
Hi,

Great lib overall, thanks for sharing your work.

I have a mapping issue when using `long` Ids. 

Here is a short console app sample that reproduces the problem: [longIdsIssue.cs](https://gist.github.com/manuel-ornato/492f206320506f9e636e8f23cc54d7b2).

I create 4 dynamic Expandos with distincts `long` Ids. There is a child table Id (`int`) too.
As you see all Ids (parents and children) are distincts, so in the end we should have 4 distinct instances of the parent class, each one being linked to  a distinct instance of the child class.

Instead, the mapping results in only 2 instances of the parent class, one containing 3 children and the other containing the last child.

The problem goes away if I change Ids to `int`.

Manuel

**Comment by manuel-ornato** [comment]
Hi,

I found the culprit in `GetIdentifierHash` (`InternalHelpers.cs`):
```
unchecked
{
    // Include identifier hashcode to avoid collisions between e.g. multiple int IDs
    identifierHash += identifierValue.GetHashCode() + identifier.GetHashCode();
}
```
One can't rely on `GetHashCode()` to uniquely identify a `long`. Actually one can't rely on an hash code to uniquely identify anything... It's just a hash, not a key.

Colllisions will happen and well... they did. In the sample I provided, 4 of my 3 long values have the same hash code. In my real database there are more than 1000 collisions actually.

I guess the fix would be to maintain your own hastable collision mecanism. That is `ìnstanceCache` should be a `Dictionary<Tuple<int, int, object>, ICollection<object>>()` and then you would have to compare the real identifiers before returning an instance.

Hope this helps.

Manuel
